Sources/KSCrashRecordingCore/KSBinaryImageCache.c

```diff
@@ -137,71 +68,55 @@ void ksbic_init(void)
 
     KSLOG_DEBUG("Initializing binary image cache");
 
-    _dyld_register_func_for_add_image(ksbic_addImageCallback);
-    _dyld_register_func_for_remove_image(ksbic_removeImageCallback);
-}
-
-// For testing purposes only. Used with extern in test files.
-void ksbic_resetCache(void)
-{
-    bool expected = true;
-    if (!atomic_compare_exchange_strong(&g_initialized, &expected, false)) {
+    struct task_dyld_info dyld_info;
+    mach_msg_type_number_t count = TASK_DYLD_INFO_COUNT;
+    kern_return_t err = task_info(mach_task_self(), TASK_DYLD_INFO, (task_info_t)&dyld_info, &count);
+    if (err != KERN_SUCCESS) {
+        KSLOG_DEBUG("Initializing binary error");
         return;
     }
-
-    pthread_rwlock_wrlock(&g_imageCacheRWLock);
-
-    for (uint32_t i = 0; i < g_cachedImageCount; i++) {
-        if (g_binaryImageCache[i].name != NULL) {
-            free((void *)g_binaryImageCache[i].name);
-            g_binaryImageCache[i].name = NULL;
-        }
-    }
-
-    memset(g_binaryImageCache, 0, sizeof(g_binaryImageCache));
-    g_cachedImageCount = 0;
-
-    pthread_rwlock_unlock(&g_imageCacheRWLock);
-}
-
-uint32_t ksbic_imageCount(void)
-{
-    uint32_t count;
-    pthread_rwlock_rdlock(&g_imageCacheRWLock);
-    count = g_cachedImageCount;
-    pthread_rwlock_unlock(&g_imageCacheRWLock);
-    return count;
+    ksbic_set_all_image_infos((struct dyld_all_image_infos *)dyld_info.all_image_info_addr);
 }
 
-const struct mach_header *ksbic_imageHeader(uint32_t index)
+const struct dyld_image_info *ksbic_beginImageAccess(uint32_t *count)
 {
-    const struct mach_header *header = NULL;
-    pthread_rwlock_rdlock(&g_imageCacheRWLock);
-    if (index < g_cachedImageCount) {
-        header = g_binaryImageCache[index].header;
+    struct dyld_all_image_infos *allInfo = ksbic_get_all_image_infos();
+    if (!allInfo) {
+        KSLOG_ERROR("Already accesing images");
```

Comment:
If the list is always stable or only additive, why do we need to guard against multiple accesses?

Reply:
Still good to make an API around it we control, what do you think?

Reply:
Yep, a wrapper API is fine, and gives us room to expand the API with search facilities. But I don't think we need `ksbic_endImageAccess` or any of the NULL stuff.